File: sdk/python/aragora/namespaces/queue.py
"""
Queue Namespace API.

Provides access to queue management endpoints for background jobs.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class QueueAPI:
    """Synchronous queue API."""

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def list_jobs(self, **params: Any) -> dict[str, Any]:
        return self._client.request("GET", "/api/queue/jobs", params=params)

    def enqueue(self, job_type: str, payload: dict[str, Any], **params: Any) -> dict[str, Any]:
        data = {"type": job_type, "payload": payload, **params}
        return self._client.request("POST", "/api/queue/jobs", json=data)

    def get_job(self, job_id: str) -> dict[str, Any]:
        return self._client.request("GET", f"/api/queue/jobs/{job_id}")

    def cancel_job(self, job_id: str) -> dict[str, Any]:
        return self._client.request("DELETE", f"/api/queue/jobs/{job_id}")

    def retry_job(self, job_id: str) -> dict[str, Any]:
        return self._client.request("POST", f"/api/queue/jobs/{job_id}/retry")

    def get_stats(self) -> dict[str, Any]:
        return self._client.request("GET", "/api/queue/stats")

    def list_workers(self) -> dict[str, Any]:
        return self._client.request("GET", "/api/queue/workers")

    def list_dlq(self, limit: int = 50) -> dict[str, Any]:
        return self._client.request("GET", "/api/queue/dlq", params={"limit": limit})

    def requeue_dlq(self, job_ids: list[str] | None = None) -> dict[str, Any]:
        payload = {"job_ids": job_ids or []}
        return self._client.request("POST", "/api/queue/dlq/requeue", json=payload)

    def cleanup(self, older_than_days: int = 7, status: list[str] | None = None) -> dict[str, Any]:
        payload = {"older_than_days": older_than_days}
        if status is not None:
            payload["status"] = status
        return self._client.request("POST", "/api/queue/cleanup", json=payload)

    def list_stale(self, threshold_minutes: int = 30) -> dict[str, Any]:
        return self._client.request(
            "GET",
            "/api/queue/stale",
            params={"threshold_minutes": threshold_minutes},
        )


class AsyncQueueAPI:
    """Asynchronous queue API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def list_jobs(self, **params: Any) -> dict[str, Any]:
        return await self._client.request("GET", "/api/queue/jobs", params=params)

    async def enqueue(
        self, job_type: str, payload: dict[str, Any], **params: Any
    ) -> dict[str, Any]:
        data = {"type": job_type, "payload": payload, **params}
        return await self._client.request("POST", "/api/queue/jobs", json=data)

    async def get_job(self, job_id: str) -> dict[str, Any]:
        return await self._client.request("GET", f"/api/queue/jobs/{job_id}")

    async def cancel_job(self, job_id: str) -> dict[str, Any]:
        return await self._client.request("DELETE", f"/api/queue/jobs/{job_id}")

    async def retry_job(self, job_id: str) -> dict[str, Any]:
        return await self._client.request("POST", f"/api/queue/jobs/{job_id}/retry")

    async def get_stats(self) -> dict[str, Any]:
        return await self._client.request("GET", "/api/queue/stats")

    async def list_workers(self) -> dict[str, Any]:
        return await self._client.request("GET", "/api/queue/workers")

    async def list_dlq(self, limit: int = 50) -> dict[str, Any]:
        return await self._client.request("GET", "/api/queue/dlq", params={"limit": limit})

    async def requeue_dlq(self, job_ids: list[str] | None = None) -> dict[str, Any]:
        payload = {"job_ids": job_ids or []}
        return await self._client.request("POST", "/api/queue/dlq/requeue", json=payload)

    async def cleanup(
        self, older_than_days: int = 7, status: list[str] | None = None
    ) -> dict[str, Any]:
        payload = {"older_than_days": older_than_days}
        if status is not None:
            payload["status"] = status
        return await self._client.request("POST", "/api/queue/cleanup", json=payload)

    async def list_stale(self, threshold_minutes: int = 30) -> dict[str, Any]:
        return await self._client.request(
            "GET",
            "/api/queue/stale",
            params={"threshold_minutes": threshold_minutes},
        )

File: sdk/typescript/src/namespaces/queue.ts
/**
 * Queue Namespace API
 *
 * Provides access to background job queues, worker status, and DLQ management.
 */

export interface QueueJob {
  job_id: string;
  type?: string;
  status?: string;
  payload?: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
  attempts?: number;
  error?: string | null;
}

export interface QueueStats {
  pending?: number;
  processing?: number;
  completed?: number;
  failed?: number;
  dead_letter?: number;
  queue_depth?: number;
}

export interface QueueWorker {
  worker_id: string;
  status: string;
  last_seen?: string;
  current_job?: string | null;
}

interface QueueClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; body?: unknown; json?: Record<string, unknown> }
  ): Promise<T>;
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
  delete<T>(path: string): Promise<T>;
}

export class QueueAPI {
  constructor(private client: QueueClientInterface) {}

  async listJobs(params?: { status?: string; type?: string; limit?: number; offset?: number }): Promise<{ jobs: QueueJob[]; total?: number }> {
    return this.client.request('GET', '/api/v1/queue/jobs', { params: params as Record<string, unknown> });
  }

  async enqueueJob(body: Record<string, unknown>): Promise<{ job_id: string }> {
    return this.client.post('/api/v1/queue/jobs', body);
  }

  async getJob(jobId: string): Promise<QueueJob> {
    return this.client.get(`/api/v1/queue/jobs/${jobId}`);
  }

  async deleteJob(jobId: string): Promise<{ deleted?: boolean }> {
    return this.client.delete(`/api/v1/queue/jobs/${jobId}`);
  }

  async retryJob(jobId: string): Promise<{ requeued?: boolean }> {
    return this.client.post(`/api/v1/queue/jobs/${jobId}/retry`);
  }

  async getStats(): Promise<QueueStats> {
    return this.client.get('/api/v1/queue/stats');
  }

  async listWorkers(): Promise<{ workers: QueueWorker[] }> {
    return this.client.get('/api/v1/queue/workers');
  }

  async listDLQ(params?: { limit?: number; offset?: number }): Promise<{ jobs: QueueJob[]; total?: number }> {
    return this.client.request('GET', '/api/v1/queue/dlq', { params: params as Record<string, unknown> });
  }

  async requeueDLQ(body?: { job_ids?: string[] }): Promise<{ requeued?: number; failed?: number }> {
    return this.client.post('/api/v1/queue/dlq/requeue', body);
  }

  async cleanup(params?: { max_age_days?: number }): Promise<{ cleaned?: number }> {
    return this.client.request('POST', '/api/v1/queue/cleanup', { params: params as Record<string, unknown> });
  }

  async listStale(params?: { stale_after_seconds?: number }): Promise<{ jobs: QueueJob[]; total?: number }> {
    return this.client.request('GET', '/api/v1/queue/stale', { params: params as Record<string, unknown> });
  }
}
